tests/rados/test_pool_osd_recovery.py

```diff
@@ -0,0 +1,348 @@
+"""
+Module to perform tier-4 tests on pools WRT OSD daemons.
+
+1. Reboot of single osd and health check
+2. Rolling Reboot OSD hosts
+3. Stopping and starting OSD daemons
+4. restart all OSD daemons belonging to single pg
+5. Removal and addition of OSD daemons
+6. Removal and addition of OSD Hosts
+
+"""
+import datetime
+import json
+import time
+
+import yaml
+
+from ceph.ceph_admin import CephAdmin
+from ceph.rados import utils
+from ceph.rados.core_workflows import RadosOrchestrator
+from ceph.rados.pool_workflows import PoolFunctions
+from ceph.utils import get_node_by_id
+from tests.ceph_installer.test_cephadm import run as add_osd
+from tests.cephadm.test_host import run as deploy_host
+from tests.rados.rados_test_util import get_device_path, wait_for_device
+from tests.rados.stretch_cluster import get_osd_details, wait_for_clean_pg_sets
+from tests.rados.test_data_migration_bw_pools import create_given_pool
+from utility.log import Log
+from utility.utils import method_should_succeed
+
+log = Log(__name__)
+
+
+def run(ceph_cluster, **kw) -> int:
+    """
+    Test to perform tier-4 tests on pools related to OSD daemons
+    Returns:
+        1 -> Fail, 0 -> Pass
+    """
+    log.info(run.__doc__)
+    config = kw["config"]
+    cephadm = CephAdmin(cluster=ceph_cluster, **config)
+    rados_obj = RadosOrchestrator(node=cephadm)
+    pool_obj = PoolFunctions(node=cephadm)
+    pool_configs = config["pool_configs"]
+    pool_configs_path = config["pool_configs_path"]
+    osd_nodes = ceph_cluster.get_nodes(role="osd")
+
+    with open(pool_configs_path, "r") as fd:
+        pool_conf_file = yaml.safe_load(fd)
+
+    pools = []
+    for i in pool_configs:
+        pool = pool_conf_file[i["type"]][i["conf"]]
+        create_given_pool(rados_obj, pool)
+        pools.append(pool["pool_name"])
+
+    log.info(f"Created {len(pools)} pools for testing. pools : {pools}")
+
+    # Checking cluster health before starting the tests
+    method_should_succeed(rados_obj.run_pool_sanity_check)
+    log.info(
+        "Completed health cehck of the cluster after test pool creation. Cluster health good!!"
+    )
+
+    log.info("\n\n---- Starting All workflows ----\n\n")
+    # rebooting one OSD and checking cluster health
+    log.info(
+        "---- Starting workflow ----\n---- 1. Reboot of single osd and health check ----"
+    )
+    for pool in pools:
+        pool_id = pool_obj.get_pool_id(pool_name=pool)
+        pgid = f"{pool_id}.0"
```

Comment:
Is it necessary to fetch pool id, as we can get the pg acting set using pool name directly

Reply:
Good Catch. Since i'm anyway always passing pool_id.0 always, It would return me the same results.

Updated the code to pass pool name directly and get the OSD list.